### PR TITLE
[13.0][MIG] website_snippet_preset

### PR DESCRIPTION
**Pull request by tarteo**

### CODE REVIEW

File: website_snippet_preset/static/src/css/website_snippet_preset.css [hunk 1]
```diff
@@ -0,0 +1,14 @@
+/* Copyright 2018 Onestein
+ * License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl). */
+
+body #oe_manipulators .oe_overlay > .oe_overlay_options {
+    width: 400px;
+}
+
+.o_snippet_preset_del {
+    position: absolute;
+    right: 0px;
```
Comment (Tardo) [inline website_snippet_preset/static/src/css/website_snippet_preset.css:10]:
```suggestion
    right: 0;
```

File: website_snippet_preset/static/src/css/website_snippet_preset.css [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+    top: 0px;
```
Comment (Tardo) [inline website_snippet_preset/static/src/css/website_snippet_preset.css:11]:
```suggestion
    top: 0;
```

File: website_snippet_preset/static/src/js/website_snippet_preset.js [hunk 3]
```diff
@@ -0,0 +1,252 @@
+/* Copyright 2018 Onestein
+ * License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl). */
+
+odoo.define('website_snippet_preset', function (require) {
+    'use strict';
+    var editor = require('web_editor.snippet.editor');
```
Comment (Tardo) [inline website_snippet_preset/static/src/js/website_snippet_preset.js:6]:
```suggestion
    const editor = require('web_editor.snippet.editor');
```

File: website_snippet_preset/static/src/js/website_snippet_preset.js [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+    var core = require('web.core');
```
Comment (Tardo) [inline website_snippet_preset/static/src/js/website_snippet_preset.js:7]:
```suggestion
    const core = require('web.core');
```

File: website_snippet_preset/static/src/js/website_snippet_preset.js [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+    var Dialog = require('web.Dialog');
```
Comment (Tardo) [inline website_snippet_preset/static/src/js/website_snippet_preset.js:8]:
```suggestion
    const Dialog = require('web.Dialog');
```

File: website_snippet_preset/static/src/js/website_snippet_preset.js [hunk 6]
```diff
@@ -0,0 +1,252 @@
...
+    var qweb = core.qweb;
```
Comment (Tardo) [inline website_snippet_preset/static/src/js/website_snippet_preset.js:9]:
```suggestion
    const qweb = core.qweb;
```

File: website_snippet_preset/static/src/js/website_snippet_preset.js [hunk 7]
```diff
@@ -0,0 +1,252 @@
...
+    var _t = core._t;
```
Comment (Tardo) [inline website_snippet_preset/static/src/js/website_snippet_preset.js:10]:
```suggestion
    const _t = core._t;
```

File: website_snippet_preset/static/src/js/website_snippet_preset.js [hunk 8]
```diff
@@ -0,0 +1,252 @@
...
+
+    var OverwriteDialog = Dialog.extend({
+        init: function (parent) {
+            this.overwrite = $.Deferred();
+            var options = {
```
Comment (Tardo) [inline website_snippet_preset/static/src/js/website_snippet_preset.js:15]:
```suggestion
            const options = {
```

File: website_snippet_preset/static/src/js/website_snippet_preset.js [hunk 9]
```diff
@@ -0,0 +1,252 @@
...
+        },
+        open: function () {
+            this._super.apply(this, arguments);
+            return this.overwrite.promise();
+        },
+    });
+
+    var EnterNameDialog = Dialog.extend({
+        init: function (parent) {
+            var options = {
```
Comment (Tardo) [inline website_snippet_preset/static/src/js/website_snippet_preset.js:48]:
```suggestion
            const options = {
```

File: website_snippet_preset/static/src/js/website_snippet_preset.js [hunk 10]
```diff
@@ -0,0 +1,252 @@
...
+        saveClicked: function () {
+            this.trigger('save', this.$('input[name="name"]').val());
+        },
+    });
+
+    var ChoiceDialog = Dialog.extend({
+        init: function (parent, snippetName) {
+            this.snippetName = snippetName;
+
+            var options = {
```
Comment (Tardo) [inline website_snippet_preset/static/src/js/website_snippet_preset.js:78]:
```suggestion
            const options = {
```

File: website_snippet_preset/static/src/js/website_snippet_preset.js [hunk 11]
```diff
@@ -0,0 +1,252 @@
...
+                        text: _t("Cancel"),
+                        classes: 'btn-default',
+                        close: true,
+                    },
+                ],
+            };
+            return this._super(parent, options);
+        },
+        start: function () {
+            var res = this._super.apply(this, arguments);
```
Comment (Tardo) [inline website_snippet_preset/static/src/js/website_snippet_preset.js:95]:
```suggestion
            const res = this._super.apply(this, arguments);
```

File: website_snippet_preset/static/src/js/website_snippet_preset.js [hunk 12]
```diff
@@ -0,0 +1,252 @@
...
+            this.$list = this.$el.find('.list-group');
+            this.loadPresets().then(function (records) {
+                this.renderPresets(records);
+            }.bind(this));
```
Comment (Tardo) [inline website_snippet_preset/static/src/js/website_snippet_preset.js:99]:
```suggestion
            this.loadPresets().then(records => {
                this.renderPresets(records);
            });
```

File: website_snippet_preset/static/src/js/website_snippet_preset.js [hunk 13]
```diff
@@ -0,0 +1,252 @@
...
+            return res;
+        },
+        itemClicked: function (arch) {
+            this.trigger('apply', arch);
+            this.close();
+        },
+        deletePreset: function (id) {
+            var self = this;
```
Comment (Tardo) [inline website_snippet_preset/static/src/js/website_snippet_preset.js:107]:
```suggestion
```

File: website_snippet_preset/static/src/js/website_snippet_preset.js [hunk 14]
```diff
@@ -0,0 +1,252 @@
...
+            return this._rpc({
+                model: 'snippet.preset',
+                method: 'unlink',
+                args: [[id]],
+            }).then(function () {
+                self.loadPresets().then(function (records) {
+                    self.renderPresets(records);
+                });
+            });
```
Comment (Tardo) [inline website_snippet_preset/static/src/js/website_snippet_preset.js:116]:
```suggestion
            }).then(() => {
                this.loadPresets().then(records => {
                    this.renderPresets(records);
                });
            });
```

File: website_snippet_preset/static/src/js/website_snippet_preset.js [hunk 15]
```diff
@@ -0,0 +1,252 @@
...
+        loadPresets: function () {
+            return this._rpc({
+                model: 'snippet.preset',
+                method: 'search_read',
+                order: 'name asc',
+                domain: [['snippet', '=', this.snippetName]],
+            });
+        },
+        renderPreset: function (record) {
+            var self = this;
```
Comment (Tardo) [inline website_snippet_preset/static/src/js/website_snippet_preset.js:127]:
```suggestion
```

File: website_snippet_preset/static/src/js/website_snippet_preset.js [hunk 16]
```diff
@@ -0,0 +1,252 @@
...
+            var $item = $(qweb
+                .render('website_snippet_preset.Item', {'record': record}));
+            $item.data('record', record);
+            $item.click(function () {
+                self.itemClicked($(this).data('record').arch);
+            });
```
Comment (Tardo) [inline website_snippet_preset/static/src/js/website_snippet_preset.js:133]:
```suggestion
            $item.click(() => {
                this.itemClicked($(this).data('record').arch);
            });
```

File: website_snippet_preset/static/src/js/website_snippet_preset.js [hunk 17]
```diff
@@ -0,0 +1,252 @@
...
+            $item.find('.badge').click(function (e) {
+                self.deletePreset($(this).closest('.list-group-item').data('record').id);
+                e.stopPropagation();
+            });
```
Comment (Tardo) [inline website_snippet_preset/static/src/js/website_snippet_preset.js:137]:
```suggestion
            $item.find('.badge').click(e => {
                this.deletePreset($(this).closest('.list-group-item').data('record').id);
                e.stopPropagation();
            });
```

File: website_snippet_preset/static/src/js/website_snippet_preset.js [hunk 18]
```diff
@@ -0,0 +1,252 @@
...
+            this.$list.append($item);
+            return $item;
+        },
+        renderPresets: function (records) {
+            this.$list.html('');
+            for (var i in records) {
```
Comment (Tardo) [inline website_snippet_preset/static/src/js/website_snippet_preset.js:143]:
```suggestion
            for (const i in records) {
```

File: website_snippet_preset/static/src/js/website_snippet_preset.js [hunk 19]
```diff
@@ -0,0 +1,252 @@
...
+    editor.Editor.include({
+        xmlDependencies: editor.Editor
+            .prototype.xmlDependencies.concat(
+                [
+                    '/website_snippet_preset/static/' +
+                    'src/xml/website_snippet_preset.xml',
+                ]
+            ),
+        _initializeOptions: function () {
+            var res = this._super.apply(this, arguments);
```
Comment (Tardo) [inline website_snippet_preset/static/src/js/website_snippet_preset.js:158]:
```suggestion
            const res = this._super.apply(this, arguments);
```

File: website_snippet_preset/static/src/js/website_snippet_preset.js [hunk 20]
```diff
@@ -0,0 +1,252 @@
...
+            var $optionsSection = _.last(this._customize$Elements);
```
Comment (Tardo) [inline website_snippet_preset/static/src/js/website_snippet_preset.js:159]:
```suggestion
            const $optionsSection = _.last(this._customize$Elements);
```

File: website_snippet_preset/static/src/js/website_snippet_preset.js [hunk 21]
```diff
@@ -0,0 +1,252 @@
...
+            var snippetName = this.$target.data().name;
```
Comment (Tardo) [inline website_snippet_preset/static/src/js/website_snippet_preset.js:160]:
```suggestion
            const snippetName = this.$target.data().name;
```

File: website_snippet_preset/static/src/js/website_snippet_preset.js [hunk 22]
```diff
@@ -0,0 +1,252 @@
...
+
+            var $presetLoad = $optionsSection.find('.o_snippet_preset_load');
```
Comment (Tardo) [inline website_snippet_preset/static/src/js/website_snippet_preset.js:162]:
```suggestion
            const $presetLoad = $optionsSection.find('.o_snippet_preset_load');
```

File: website_snippet_preset/static/src/js/website_snippet_preset.js [hunk 23]
```diff
@@ -0,0 +1,252 @@
...
+            $presetLoad.data('snippetName', snippetName);
+            $presetLoad.click(this.loadPreset.bind(this));
+
+            var $presetSave = $optionsSection.find('.o_snippet_preset_save');
```
Comment (Tardo) [inline website_snippet_preset/static/src/js/website_snippet_preset.js:166]:
```suggestion
            const $presetSave = $optionsSection.find('.o_snippet_preset_save');
```

File: website_snippet_preset/static/src/js/website_snippet_preset.js [hunk 24]
```diff
@@ -0,0 +1,252 @@
...
+            $presetSave.click(this.saveCurrentPreset.bind(this));
+
+            if (!snippetName) {
+                $presetLoad.addClass('d-none');
+                $presetSave.addClass('d-none');
+            }
+            return res;
+        },
+        loadPreset: function (event) {
+            var $target = $(event.target);
```
Comment (Tardo) [inline website_snippet_preset/static/src/js/website_snippet_preset.js:177]:
```suggestion
            const $target = $(event.target);
```

File: website_snippet_preset/static/src/js/website_snippet_preset.js [hunk 25]
```diff
@@ -0,0 +1,252 @@
...
+            var snippetName = $target.data('snippetName');
```
Comment (Tardo) [inline website_snippet_preset/static/src/js/website_snippet_preset.js:178]:
```suggestion
            const snippetName = $target.data('snippetName');
```

File: website_snippet_preset/static/src/js/website_snippet_preset.js [hunk 26]
```diff
@@ -0,0 +1,252 @@
...
+
+            var self = this;
```
Comment (Tardo) [inline website_snippet_preset/static/src/js/website_snippet_preset.js:180]:
```suggestion
```

File: website_snippet_preset/static/src/js/website_snippet_preset.js [hunk 27]
```diff
@@ -0,0 +1,252 @@
...
+            var dialog = new ChoiceDialog(this, snippetName);
```
Comment (Tardo) [inline website_snippet_preset/static/src/js/website_snippet_preset.js:181]:
```suggestion
            const dialog = new ChoiceDialog(this, snippetName);
```

File: website_snippet_preset/static/src/js/website_snippet_preset.js [hunk 28]
```diff
@@ -0,0 +1,252 @@
...
+            dialog.on('apply', this, function (arch) {
+                self.applyPreset(arch);
+            });
```
Comment (Tardo) [inline website_snippet_preset/static/src/js/website_snippet_preset.js:184]:
```suggestion
            dialog.on('apply', this, arch => {
                this.applyPreset(arch);
            });
```

File: website_snippet_preset/templates/assets.xml [hunk 29]
```diff
@@ -0,0 +1,13 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!-- Copyright 2018 Onestein
+     License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl). -->
+
+<odoo>
+    <template id="assets_editor" inherit_id="website.assets_editor">
+        <xpath expr="." position="inside">
+            <script src="/website_snippet_preset/static/src/js/website_snippet_preset.js"></script>
```
Comment (Tardo) [inline website_snippet_preset/templates/assets.xml:8]:
```suggestion
            <script type="text/javascript" src="/website_snippet_preset/static/src/js/website_snippet_preset.js"></script>
```

File: website_snippet_preset/static/src/js/website_snippet_preset.js [hunk 30]
```diff
@@ -0,0 +1,252 @@
...
+            dialog.open();
+        },
+        saveCurrentPreset: function (event) {
+            var arch = this.$target[0].outerHTML;
```
Comment (Tardo) [inline website_snippet_preset/static/src/js/website_snippet_preset.js:188]:
```suggestion
            const arch = this.$target[0].outerHTML;
```

File: website_snippet_preset/static/src/js/website_snippet_preset.js [hunk 31]
```diff
@@ -0,0 +1,252 @@
...
+            var snippet = $(event.target).data('snippetName');
```
Comment (Tardo) [inline website_snippet_preset/static/src/js/website_snippet_preset.js:189]:
```suggestion
            const snippet = $(event.target).data('snippetName');
```

File: website_snippet_preset/static/src/js/website_snippet_preset.js [hunk 32]
```diff
@@ -0,0 +1,252 @@
...
+
+            var self = this;
```
Comment (Tardo) [inline website_snippet_preset/static/src/js/website_snippet_preset.js:191]:
```suggestion
```

File: website_snippet_preset/static/src/js/website_snippet_preset.js [hunk 33]
```diff
@@ -0,0 +1,252 @@
...
+            var dialog = new EnterNameDialog(this);
```
Comment (Tardo) [inline website_snippet_preset/static/src/js/website_snippet_preset.js:192]:
```suggestion
            const dialog = new EnterNameDialog(this);
```

File: website_snippet_preset/static/src/js/website_snippet_preset.js [hunk 34]
```diff
@@ -0,0 +1,252 @@
...
+                        self.savePreset(snippet, name, arch, exists);
+                        dialog.close();
+                    };
+                    if (exists) {
+                        new OverwriteDialog(dialog).open().done(save);
+                    } else {
+                        save();
+                    }
+                });
+            });
```
Comment (Tardo) [inline website_snippet_preset/static/src/js/website_snippet_preset.js:205]:
```suggestion
            dialog.on('save', this, name => {
                this.presetExists(snippet, name).done(exists => {
                    const save = () => {
                        this.savePreset(snippet, name, arch, exists);
                        dialog.close();
                    };
                    if (exists) {
                        new OverwriteDialog(dialog).open().done(save);
                    } else {
                        save();
                    }
                });
            });
```

File: website_snippet_preset/static/src/js/website_snippet_preset.js [hunk 35]
```diff
@@ -0,0 +1,252 @@
...
+            dialog.open();
+        },
+        presetExists: function (snippet, name) {
+            var get_done = $.Deferred();
```
Comment (Tardo) [inline website_snippet_preset/static/src/js/website_snippet_preset.js:209]:
```suggestion
            const get_done = $.Deferred();
```

File: website_snippet_preset/static/src/js/website_snippet_preset.js [hunk 36]
```diff
@@ -0,0 +1,252 @@
...
+                save = this._rpc({
+                    model: 'snippet.preset',
+                    method: 'create',
+                    args: [{arch: arch, name: name, snippet: snippet}],
+                });
+            }
+            return save;
+        },
+        applyPreset: function (arch) {
+            var $arch = $(arch);
```
Comment (Tardo) [inline website_snippet_preset/static/src/js/website_snippet_preset.js:239]:
```suggestion
            const $arch = $(arch);
```

File: website_snippet_preset/static/src/js/website_snippet_preset.js [hunk 37]
```diff
@@ -0,0 +1,252 @@
...
+                method: 'search_read',
+                domain: [['snippet', '=', snippet], ['name', '=', name]],
+            }).then(function (records) {
+                get_done
+                    .resolve(records.length === 0 ? false : records[0].id);
+            });
+            return get_done.promise();
+        },
+        savePreset: function (snippet, name, arch, id) {
+            var self = this;
```
Comment (Tardo) [inline website_snippet_preset/static/src/js/website_snippet_preset.js:221]:
```suggestion
```

File: website_snippet_preset/static/src/js/website_snippet_preset.js [hunk 38]
```diff
@@ -0,0 +1,252 @@
...
+            var save = null;
```
Comment (Tardo) [inline website_snippet_preset/static/src/js/website_snippet_preset.js:222]:
```suggestion
            let save = null;
```

**Review by Tardo — CHANGES_REQUESTED** [review]
Thanks! please, use ES6

**Comment by Max-282** [comment]
Hey @tarteo ! Thanks, it work very well. We need the V13 for our migration and it seems to be practically done. I was wondering if the project was at a standstill or you thought to finish it, we can take it over if necessary.

**Pull request closed without merging**